core/services/head_tracker.go

```diff
@@ -277,32 +282,119 @@ func (ht *HeadTracker) handleNewHead(bh gethTypes.Header) error {
 	}
 	if head.Number == prevHead.Number {
 		if head.Hash != prevHead.Hash {
-			logger.Debugf("duplicate blocks at height %v. Got block hash %s but already saw block hash %s", head.Number, head.Hash.Hex(), ht.highestSeenHead.Hash.Hex())
+			logger.Debugw("HeadTracker: got duplicate head", "blockNum", head.Number, "gotHead", head.Hash.Hex(), "highestSeenHead", ht.highestSeenHead.Hash.Hex())
 		} else {
-			logger.Debugf("head with hash %s was already in the database", head.Hash.Hex())
+			logger.Debugw("HeadTracker: head already in the database", "gotHead", head.Hash.Hex())
 		}
 	} else {
-		logger.Debugf("received out of order head %s with number %v. Latest head is at %v", head.Hash.Hex(), head.Number, ht.highestSeenHead.Number)
+		logger.Debugw("HeadTracker: got out of order head", "blockNum", head.Number, "gotHead", head.Hash.Hex(), "highestSeenHead", ht.highestSeenHead.Number)
 	}
 	return nil
 }
 
 func (ht *HeadTracker) handleNewHighestHead(head models.Head) error {
-	headWithChain, err := ht.store.Chain(head.Hash, ht.store.Config.EthFinalityDepth())
+	// NOTE: We must set a hard time limit on this, backfilling heads should
+	// not block the head tracker
+	ctx, cancel := context.WithTimeout(context.Background(), backfillTimeBudget)
+	defer cancel()
+
+	headWithChain, err := ht.GetChainWithBackfill(ctx, head, ht.store.Config.EthFinalityDepth())
 	if err != nil {
 		return err
 	}
-	if headWithChain == nil {
-		return fmt.Errorf("invariant violation: head with block hash %s was missing", head.Hash)
+
+	ht.onNewLongestChain(headWithChain)
+	return nil
+}
+
+// GetChainWithBackfill returns a chain of the given length, backfilling any
+// heads that may be missing from the database
+func (ht *HeadTracker) GetChainWithBackfill(ctx context.Context, head models.Head, depth uint) (models.Head, error) {
+	var err error
+	head, err = ht.store.Chain(head.Hash, depth)
+	if err != nil {
+		return head, errors.Wrap(err, "GetChainWithBackfill failed fetching chain")
+	}
+	if uint(head.ChainLength()) >= depth {
+		return head, nil
+	}
+	baseHeight := head.Number - int64(depth-1)
+	if baseHeight < 0 {
+		baseHeight = 0
 	}
-	ht.onNewLongestChain(*headWithChain)
+
+	if err := ht.backfill(ctx, head.EarliestInChain(), baseHeight); err != nil {
+		return head, errors.Wrap(err, "GetChainWithBackfill failed backfilling")
+	}
+	return ht.store.Chain(head.Hash, depth)
+}
+
+// backfill fetches all missing heads up until the base height
+func (ht *HeadTracker) backfill(ctx context.Context, head models.Head, baseHeight int64) error {
```

Comment:
can we make baseHeight a uint to enforce baseHeight > 0?

Reply:
mm, wouldn't that result in overflow/panic if the result would otherwise be negative?

Reply:
It means I would have to put explicit type casts everywhere since we compare it a lot to head.Number which is `int64`